internal/configuration/provider.go

```diff
@@ -0,0 +1,103 @@
+package configuration
+
+import (
+	"fmt"
+
+	"github.com/knadh/koanf"
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/authelia/authelia/internal/configuration/schema"
+	"github.com/authelia/authelia/internal/configuration/validator"
+)
+
+// LoadSources is a variadic function that takes types that implement the Source interface.
+func (p *Provider) LoadSources(sources ...Source) (errs []error) {
+	for _, source := range sources {
+		err := source.Load()
+		if err != nil {
+			p.Validator.Push(fmt.Errorf("failed to load configuration from %s source: %+v", source.Name(), err))
+
+			continue
+		}
+
+		err = source.Merge(p.Koanf)
+		if err != nil {
+			p.Validator.Push(fmt.Errorf("failed to merge configuration from %s source: %+v", source.Name(), err))
+
+			continue
+		}
+
+		val := source.Validator()
+
+		if val != nil {
+			for _, err := range val.Errors() {
+				p.Validator.Push(fmt.Errorf("%s: %+v", source.Name(), err))
+			}
+		}
+	}
+
+	return p.Validator.Errors()
+}
+
+// Unmarshal the koanf.Koanf.
+func (p *Provider) Unmarshal() (warns []error, errs []error) {
+	if p.validation {
+		validator.ValidateKeys(p.Koanf.Keys(), p.Validator)
+	}
+
+	p.Configuration = &schema.Configuration{}
+
+	err := p.unmarshal("", p.Configuration)
+	if err != nil {
+		p.Validator.Push(fmt.Errorf("error occurred during unmarshalling configuration: %+v", err))
+	} else if p.validation {
+		validator.ValidateConfiguration(p.Configuration, p.Validator)
+	}
+
+	return p.Validator.Warnings(), p.Validator.Errors()
+}
+
+// SetValidation changes the validation state of the provider. Disabling it allows manual checking.
+func (p *Provider) SetValidation(validation bool) {
+	p.validation = validation
+}
+
+func (p *Provider) unmarshal(path string, o interface{}) (err error) {
+	c := koanf.UnmarshalConf{
+		DecoderConfig: &mapstructure.DecoderConfig{
+			DecodeHook: mapstructure.ComposeDecodeHookFunc(
+				mapstructure.StringToTimeDurationHookFunc(),
+				mapstructure.StringToSliceHookFunc(","),
+			),
+			Metadata:         nil,
+			Result:           o,
+			WeaklyTypedInput: true,
+		},
+	}
+
+	return p.Koanf.UnmarshalWithConf(path, o, c)
+}
+
+var provider *Provider
```

Comment:
I usually tend to avoid using global variables because it encourages developers to do side effects in functions which is harder to detect because it's not expressed in the function signatures and it's often harder to test in big code bases because we don't where side effects can come from. The best use case I can think of for singleton is a resource manager but here we don't need it. Let's keep the code more functional if you will.

Reply:
Does this look overall more like what you had in mind? https://gist.github.com/james-d-elliott/8228625532c803d59bf460aa40d86ad3

Reply:
@clems4ever